### PR TITLE
WIP: migrate int tests from bats to python

### PR DESCRIPTION
**Pull request by mkinney**
Working on #287 

Opening PR early so you have visibility to changes and can provide input/feedback, if you see anything you do not like.

Basically trying to pull the integration tests into pytest and clean out the examples dir a bit.

I wanted to add `pytest-xdist` but it does not play well.

### CODE REVIEW

**Comment by mkinney** [comment]
This what it looks like so far:

```
mikekinney@sweet mkinney_pyinfra % pytest -vv -l --forked -m'int'
============================================================== test session starts ==============================================================
platform darwin -- Python 3.7.6, pytest-4.6.6, py-1.8.1, pluggy-0.13.1 -- /Users/mikekinney/python/mkinney_pyinfra/venv/bin/python3.7
cachedir: .pytest_cache
rootdir: /Users/mikekinney/python/mkinney_pyinfra, inifile: pytest.ini
plugins: forked-1.1.3, cov-2.8.1
collected 565 items / 554 deselected / 11 selected

tests/int/test_int_docker.py::test_int_docker_install_package_ubuntu PASSED                                                               [  9%]
tests/int/test_int_docker.py::test_int_docker_apk_on_alpine PASSED                                                                        [ 18%]
tests/int/test_int_docker.py::test_int_docker_apt_and_npm_on_ubuntu PASSED                                                                [ 27%]
tests/int/test_int_docker.py::test_int_docker_git_gem_and_pip_on_mult PASSED                                                              [ 36%]
tests/int/test_int_docker.py::test_int_docker_pacman_on_arch PASSED                                                                       [ 45%]
tests/int/test_int_docker.py::test_int_docker_files_dnf_server_on_centos PASSED                                                           [ 54%]
tests/int/test_int_docker.py::test_int_docker_yum_on_centos PASSED                                                                        [ 63%]
tests/int/test_int_docker.py::test_int_docker_adhoc_apt_packages PASSED                                                                   [ 72%]
tests/int/test_int_docker.py::test_int_docker_adhoc_fact_os PASSED                                                                        [ 81%]
tests/int/test_int_docker.py::test_int_docker_adhoc_all_facts SKIPPED                                                                     [ 90%]
tests/int/test_int_vagrant.py::test_int_vagrant PASSED                                                                                    [100%]

=============================================================== warnings summary ================================================================

<snipped the warnings>


====================================== 10 passed, 1 skipped, 554 deselected, 12 warnings in 695.63 seconds ======================================
```

**Comment by Fizzadar** [comment]
Tested and works a charm! Wondering how/if there's a way to run these in travis.

@mkinney I'm good to merge this if you're happy?

**Comment by mkinney** [comment]
Need your help.

`E        +  where 1 = CompletedProcess(args='pyinfra @docker/ubuntu apt.py npm.py', returncode=1, stdout=b'--> Loading config...\n--> Loadin...x_command(command, **command_kwargs)\nTypeError: make_unix_command() got an unexpected keyword argument \'get_pty\'\n').returncode`

Can you get this working locally?
`pytest -m'int' tests/int/test_int_docker.py`

**Comment by mkinney** [comment]
@Fizzadar Can you see comment above? (not sure if I need to tag you or not for you to see it)

**Comment by Fizzadar** [comment]
@mkinney yep got it - think this is an edge case with the `0.14.2` changes, will check and fix that!

**Comment by mkinney** [comment]
I thought about it later...

`pytest --cov -m"int or not int and not vagrant"` could be re-written as:
`pytest --cov -m"not vagrant"` lols

**Comment by Fizzadar** [comment]
The Docker issue should now be fixed by https://github.com/Fizzadar/pyinfra/commit/0b652d6eb42f089cf121f36d1a6193279b5dbc69!

File: tests/int/conftest.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+class Helpers:
+    @staticmethod
+    def run(command, expected_stdout_lines=None, expected_stderr_lines=None,
+            cwd='examples'):
+        '''Helper function to eliminate pytest code.'''
+        if expected_stdout_lines is None:
+            expected_stdout_lines = ['Connected', 'Starting operation', 'Errors: 0']
+        if expected_stderr_lines is None:
+            expected_stderr_lines = ['']
+        results = subprocess.run(command, shell=True, cwd=cwd, capture_output=True)
```
Comment (Fizzadar) [inline tests/int/conftest.py]: The final problem is this line - `subprocess.run` doesn't exist in Py2 :(

You should be able to use popen:

```py
process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
process.wait()  # wait until we have a returncode
```

Comment (mkinney) [reply tests/int/conftest.py]: Ok

At some point, consider deprecating v2.7.

**Comment by mkinney** [comment]
All is looking good. Ok to merge if you think it is ready/looks good.

**Pull request merged**